### ISSUE TITLE
Vectorize predict method

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#511]

**Comment by khalibartan** [comment]
can you elaborate it a bit more. I'll will like to take up this issue.

**Comment by ankurankan** [comment]
@khalibartan In the predict method, right now we loop over each data point and then call the query method. We want to remove this for loop and find a more efficient way which would directly compute all the values.

**Comment by khalibartan** [comment]
which particular script is that. Or is this problem with all the scripts?

**Comment by khalibartan** [comment]
Is this in relation with #473

**Comment by ankurankan** [comment]
@khalibartan No.

**Comment by kshitij10496** [comment]
@ankurankan Are we looking to change this for a performance boost or to give the module a more pythonic look ?

**Comment by ankurankan** [comment]
@kshitij10496 For performance boost

**Comment by khalibartan** [comment]
@ankurankan Are we iterating over numpy arrays, because that is a very costly process to do, even costlier than list iteration .

**Comment by ankurankan** [comment]
@khalibartan Depends on what we are getting in the argument.

**Comment by khalibartan** [comment]
@ankurankan Are you referencing the models modules here. If not then can reference the modules which you are talking about. I would like to take up this issue.

**Comment by shwetamore1295** [comment]
Is this issue resolved? If not, I would like to take this up.

**Comment by ankurankan** [comment]
@shwetamore1295 You can go ahead with this but will have to do some research on how to do this.

**Comment by MikeDT** [comment]
Assuming no progress on this?  The pgmpy.models.BayesianModel predict function looks to be very slow.  I'm no Python expert, but my suspicion is that the following block is the culprit

    `   model_inference = VariableElimination(self)
        for index, data_point in data.iterrows():
            states_dict = model_inference.map_query(variables=missing_variables, evidence=data_point.to_dict())
            for k, v in states_dict.items():
                pred_values[k].append(v)
        return pd.DataFrame(pred_values, index=data.index)`

As an aside, would suggest nudging towards a more sklearn type interface for this, as for a lt of people this is the primary use case.  As a corollary, the closer the interface is to sklearn, the greater uptake.  

I'll have a crack at optimising :)

**Comment by ankurankan** [comment]
@MikeDT Yeah, I don't think there's been any progress on this. And you are right, it's slow because of that block as we are iterating over the data points. And I agree with you on having a sklearn kind of API, and we have been trying to keep it like that but in some cases it's not possible for us.

**Comment by MikeDT** [comment]
Ha, good to know I'm not a total noob :).  If I make any progress I'll be sure to let you know.  The rest of the code is pretty impressive performance wise, so there's a high standard...

**Comment by zachguo** [comment]
> ```
> `   model_inference = VariableElimination(self)
>     for index, data_point in data.iterrows():
>         states_dict = model_inference.map_query(variables=missing_variables, evidence=data_point.to_dict())
>         for k, v in states_dict.items():
>             pred_values[k].append(v)
>     return pd.DataFrame(pred_values, index=data.index)`
> ```

For this part, shall we change it to
```py
model_inference = VariableElimination(self)
return data.apply(
    lambda row: model_inference.map_query(variables=missing_variables, evidence=row.to_dict()),
    axis=1
)
```

**Comment by ankurankan** [comment]
@zachguo Not really, because its going to call the map query method individually for each data point. I was thinking of having some kind of way to directly work with the whole matrix of data (with matrix operations) instead of iterating over it. But I am not even sure if something like that would be possible because the computation depends on the data.
